docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,388 @@
+
+### Dependency Watchdog
+
+Currently, the [`dependency-watchdog`](https://github.com/gardener/dependency-watchdog) monitors the leases of the nodes in the shoot cluster. If nodes more than a configured threshold are unhealthy, it scales down the `machine-controller-manager`, `kube-controller-manager`, and `cluster-autoscaler` to prevent a meltdown. During an ongoing in-place update, DWD can make use of the labels added by `machine-controller-manager` during the `in-place` update to exclude nodes undergoing updates from its health check.
+
+### Extensions
+
+#### OS extensions
+
+The OS extension that wants to support in-place updates should add an "update-script" to the `.spec.files[]` of the OSC with a pre-defined path that will be executed by the `gardener-node-agent` to perform the update. This script should trigger the OS-provided update tool/utility to update the OS.
+
+### Failures and Recovery Strategy
+
+In-place updates may fail because of several factors ranging from network issues to compatibility problems. This section outlines the failure scenarios, error classifications, and the specific roles `gardener-node-agent` and `machine-controller-manager` play in failure handling and recovery to ensure minimal disruption.
+
+#### Types of Failures During Update
+
+##### Drain Timeout/Failure
+
+- A node is marked as `ready-for-update` only after a successful drain.
+- We use the existing [drain logic](https://github.com/gardener/machine-controller-manager/tree/v0.54.0/pkg/util/provider/drain) provided by the `machine-controller-manager`.
+- If a drain operation times out or fails, the node will be forcefully drained.
+
+##### Kubernetes Update Failures
+
+- **Rollback on Failure:** If a Kubernetes version update fails, `gardener-node-agent` will roll back the `kubelet` to the version specified in the last applied `OperatingSystemConfig`. This ensures that the node remains operational.
+  - **Update Halt on Failure:** `gardener-node-agent` will label the node with an `update-failed` label. This label signals `machine-controller-manager` to stop further updates.
+
+##### OS Update Failures
+
+- **Before rebooting for update** - Update cannot start if the machine image version is unavailable, network issues prevent the image pull, or disk storage is insufficient (although proper garbage collection should prevent this); classified as retriable or non-retriable.
+- **Failed to boot the new OS** - If the node fails to boot into the new OS version and reverts to the previous version, `gardener-node-agent` will annotate it to mark an update attempt, but the root cause may remain unlogged as filesystems mount later in the booting process; classified as non-retriable.
+- **Total failure** - In cases where the system fails to boot into either of the OS versions, the update attempt will timeout and be marked as failed; classified as non-retriable.
```

Comment:
Who will mark the node as failed in this case (failure to boot either OS version)?
The diagram suggests the GNA to label the node with `update-failed`, however if the OS cannot boot, GNA will certainly not run.

Reply:
It would be MCM adding this label, this is mentioned under MCM section - https://github.com/gardener/gardener/blob/1e5bcac44c77df3efafdf59aa93aa71e64debe8e/docs/proposals/31-inplace-node-update.md?plain=1#L351